packages/server-renderer/src/renderToStream.ts

```diff
@@ -0,0 +1,476 @@
+import {
+  App,
+  Component,
+  ComponentInternalInstance,
+  VNode,
+  VNodeArrayChildren,
+  createVNode,
+  Text,
+  Comment,
+  Static,
+  Fragment,
+  ssrUtils,
+  Slots,
+  Slot,
+  createApp,
+  ssrContextKey,
+  warn,
+  DirectiveBinding,
+  VNodeProps,
+  mergeProps
+} from 'vue'
+import {
+  ShapeFlags,
+  isString,
+  isPromise,
+  isFunction,
+  isVoidTag,
+  escapeHtml,
+  NO,
+  generateCodeFrame,
+  escapeHtmlComment
+} from '@vue/shared'
+import { compile } from '@vue/compiler-ssr'
+import { ssrRenderAttrs } from './helpers/ssrRenderAttrs'
+import { CompilerError } from '@vue/compiler-dom'
+import { Readable } from 'stream'
+
+const {
+  isVNode,
+  createComponentInstance,
+  setCurrentRenderingInstance,
+  setupComponent,
+  renderComponentRoot,
+  normalizeVNode,
+  normalizeSuspenseChildren
+} = ssrUtils
+
+// Each component has a buffer array.
+// A buffer array can contain one of the following:
+// - plain string
+// - A resolved buffer (recursive arrays of strings that can be unrolled
+//   synchronously)
+// - An async buffer (a Promise that resolves to a resolved buffer)
+export type SSRBufferItem = string | SSRBuffer | Promise<SSRBuffer>
+export type SSRBuffer = SSRBufferItem[]
+
+export type PushFn = (item: SSRBufferItem) => void
+
+export type Props = Record<string, unknown>
+
+export type SSRContext = {
+  [key: string]: any
+  teleports?: Record<string, string>
+  __teleportBuffers?: Record<string, SSRBuffer>
+}
+
+export type SSRSlots = Record<string, SSRSlot>
+
+export type SSRSlot = (
+  props: Props,
+  push: PushFn,
+  parentComponent: ComponentInternalInstance | null,
+  scopeId: string | null
+) => void
+
+export function createBuffer() {
+  let appendable = false
+  const buffer: SSRBuffer = []
+  return {
+    getBuffer(): SSRBuffer {
+      // Return static buffer and await on items during unroll stage
+      return buffer as SSRBuffer
+    },
+    push(item: SSRBufferItem) {
+      const isStringItem = isString(item)
+      if (appendable && isStringItem) {
+        buffer[buffer.length - 1] += item as string
+      } else {
+        buffer.push(item)
+      }
+      appendable = isStringItem
+    }
+  }
+}
+
+async function unrollBuffer(
+  buffer: SSRBuffer,
+  stream: Readable
+): Promise<void> {
+  try {
+    for await (const item of buffer) {
```

Comment:
I think it's better to only `await` when the item is a Promise - this would save a microtask tick on sync buffer entries.

Reply:
Should be fine now I think